### ISSUE TITLE
Suggest to replace type parameter with impl

### ISSUE THREAD

**Issue by vigna** [rust-lang/rust-clippy#11297]
### What it does

`fn  f<A: Sized>(x: A)` can be replaced by `fn(x: impl Sized)`, which is more readable and 99.9% sufficient for the purpose of specifying a parameter implementing a trait.

### Advantage

More readable code (type is just after the variabile, no eyeballing between type parameters and argument types).

It is useful to have it as a lint because sometimes changes in the code make it possible to move a type parameter to impl, but the programmer might miss the opportunity.

### Drawbacks

None, except personal preference (some people might not like impl). In the case of return types impl has consequences (not being able to choose the return type) but in the case of function arguments this is not an issue.

### Example

```rust
<code>
fn f<A: Sized>(a: A){
    
}

```

Could be written as:

```rust
<code>
fn f(a: impl Sized){
    
}
```

**Comment by y21** [comment]
A drawback for impl trait in argument position specifically is that it makes it impossible to turbofish the function call at call sites (e.g. `f::<i32>` is no longer allowed), so this should probably also check that there aren't any paths to such a function with generic args specified

**Comment by vigna** [comment]
True, but you are passing a parameter—you can apply an `as i32` to the parameter obtaining the same effect. I agree that the lint should warn you of turbofish usage, but I don't see this as a drawback to use impl; rather, you have to use a different syntax to force the type, and certainly some additional editing must be done. I updated the issue.

**Comment by y21** [comment]
Fair point, it's indeed not really an issue most of the time, but sometimes it's not as trivial as simply adding an `as` cast and you need to move it into its own binding just to be able to annotate the types. In those cases it would be very convenient to be able to help type inference using turbofish ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4092171e32daf3e69a9806b2c1c18962), `as`-casting doesn't work here, `f::<i32>` would be convenient, but that's not possible due to `impl Sized`).

I realize I'm being pedantic here though, otherwise it sounds like a neat lint idea.
This is proably as you said less of an issue with impl trait in argument position and more of an issue with (or lack thereof) being able to annotate types within an expression (the type_ascription feature would've probably helped here, but that doesn't exist anymore)

**Comment by vigna** [comment]
You're just being helpful—I'm still a newbie and I can't see all ramifications. Note that however this works:
```
macro_rules! default {
  () => { Default::default() }
}

fn f(_: impl Sized) {}

fn main() {
    let v: i32 = default!();
    f(v); //~ ERROR: type annotations needed
}
```
So you're not losing functionality.

**Comment by y21** [comment]
Right, you can get around it by putting it in a variable and annotate it this way, I just wanted to say that it would be more convenient in that example if you didn't need to do this and you could directly say `f::<i32>(default!())`, like [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1c846c916b0803334430c0c604f6a688) ^^
But you're right, this is just an inconvenience and it can still work, it just requires more changes at call site.

What I originally meant was that, since this can break code, clippy should be "careful" with the suggestion it gives.
There are different kinds of "applicabilities" for suggestions, `MachineApplicable` is supposed to compile after applying suggestions (`cargo clippy --fix` will auto-apply these suggestions and would break if it didn't compile).
So if it were to give such a machine applicable suggestion for this, it should make sure that it doesn't do so for functions that are being called with turbofish. But it could also just ignore this case and not emit a machine applicable suggestion
